File: scripting-langs/python/game.py
# Guessing game in Python 3.


import random


class GuessingGame:
    """
    Guessing game structure.
    """

    def __init__(self):
        self.magic = random.randint(1, 100)
        self.tries = []
        self.win = False

    def check(self, guess):
        """
        Check a guess, shows if it is too large / small, or correct, and
        record that try.
        Returns true if correct, otherwise false.
        """
        self.tries.append(guess)
        if guess == self.magic:
            print("Correct!")
            return True
        else:
            if guess < self.magic:
                print("Too small!")
            elif guess > self.magic:
                print("Too large!")
            return False

    def stats(self):
        """
        Show game stats & reset game status.
        """
        print("<Game stats>")
        print("  You tried: ", end='')
        for num in self.tries:
            print(num, end=' ')
        print()
        print("  In total " + str(len(self.tries)) + " tries used.")

        del self.tries[:]
        self.win = False

    def play(self):
        """
        Plays an guessing interactively in the shell.
        """
        print("Welcome to Guanzhou's guessing game!")

        while True:
            ipt = input("Enter an int ∈ [1, 100], or 'q' to quit: ").strip()
            if len(ipt) == 0:   # Empty input.
                continue
            if ipt[0] == 'q':   # Quit signal.
                break
            try:
                guess = int(ipt)
                if not 1 <= guess <= 100:
                    print("WARN: valid input ∈ [1, 100]. Try again...")
                else:
                    print("  Your guess is... ", end='')
                    if self.check(guess):
                        self.win = True
                        break
            except ValueError:
                print("WARN: not a valid integer. Try again...")

        if self.win:
            print("You win! Congrats ;)")
        else:
            print("Sad to see you go :(")
        self.stats()


if __name__ == "__main__":
    game = GuessingGame()
    game.play()

File: scripting-langs/julia/game.jl
# Guessing game in Julia 1.2.
# Let's call the magic number λ, and a guess γ.


"Showing Julia's functional syntax & multi-dispatch."
equal(λ, x) = λ == x
equal(λ) = x -> equal(λ, x)
# @show equal(37)


"Check a guess against a magic number, and record this try."
function check(λ, γ, tries)
    push!(tries, γ)
    if equal(λ)(γ)
        println("Correct!")
        return true
    else
        if γ < λ
            println("Too small!")
        elseif γ > λ
            println("Too large!")
        end
        return false
    end
end


"Show game stats."
function stats(tries)
    println("<Game stats>")
    print("  You tried: ")
    for num in tries
        print("$num ")
    end
    println()
    println("  In total "*string(length(tries))*" tries used.")
end


"Game logic."
function play()
    println("Welcome to Guanzhou's guessing game!")

    # Generate random number.
    λ = rand(1:100)

    # Guessing loop.
    win = false
    tries = []
    while true
        print("Enter an int ∈ [1, 100], or 'q' to quit: ")
        ipt = strip(readline())
        if length(ipt) == 0     # Empty input.
            continue
        end
        if ipt[1] == 'q'        # Quit signal.
            break
        end
        if (ipt[1] < '0' || ipt[1] > '9') && ipt[1] != '-'
            println("WARN: not a valid integer. Try again...")
        else
            γ = parse(Int, ipt)
            if !(1 <= γ <= 100)
                println("WARN: valid input ∈ [1, 100]. Try again...")
            else
                print("  You guess is... ")
                if check(λ, γ, tries)
                    win = true
                    break
                end
            end
        end
    end

    # Ending message.
    if win
        println("You win! Congrats ;)")
    else
        println("Sad to see you go :(")
    end
    stats(tries)
end


play();
